### ISSUE TITLE
[$250] [Image] Implement the ability to copy images to the clipboard

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#4629]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

Coming from https://github.com/Expensify/App/issues/3313

## Action Performed:
Right-click or long-press on an image in a chat.

## Expected Result:
1. A context menu appears, displaying a single option: `Copy image to clipboard`
1. Clicking on this option should copy the image to the clipboard, where it can be pasted somewhere else.

**Note**: the solution implemented must be an offline-first solution. Meaning the user should be able to copy the image to their clipboard, whether or not they have internet connection. 

## Actual Result:
1. The normal context menu appears.
1. Clicking on `Copy to clipboard` copies the HTML image tag to the clipboard, not the binary image itself.

## Workaround:
Screenshot the image? Also you can click on it and download it.

## Platform:
- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** `1.0.85-3`
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:**

Upwork job: https://www.upwork.com/jobs/~012fd9a807faaf42aa
[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01b3973ec74ce159e9</li>
        <li>Upwork Job ID: 1782953316490067968</li>
        <li>Last Price Increase: 2024-05-01</li>
    </ul>
</details>

**Comment by roryabraham** [comment]
Okay, there was a fair deal of discussion over in https://github.com/Expensify/App/issues/3313 and other issues, but the task at hand here is to:

1. (internal) Create an `Expensify/` fork of https://github.com/react-native-clipboard/clipboard
1. Create a separate fork of https://github.com/react-native-clipboard/clipboard for local testing.
1. Provide a complete implementation of [this issue](https://github.com/react-native-clipboard/clipboard/issues/6) in your separate fork.
1. Point this project and your fork for local testing.
1. Add a new [context menu type](https://github.com/Expensify/App/blob/0b17d56c6c2dbfac115ced3bc1c0b4a6a80bcf06/src/pages/home/report/ContextMenu/ContextMenuActions.js#L25) for images, display it when images are right-clicked or long-pressed. [This PR](https://github.com/Expensify/App/pull/3931) might serve as a useful guide.
1. Once that's working, submit a PR from your fork of `react-native-clipboard` to the `Expensify/react-native-clipboard` fork.
1. Also submit a PR from your fork of `react-native-clipboard` to the upstream project.
1. Create a PR in this repo with your implementation from step `5` above, but point the project at the `Expensify/react-native-clipboard` fork instead of your personal one. (should be on HOLD until the PR from step `6` is merged.
1. Lastly, if/when your PR from step 7 is merged by the project maintainers, create another PR to point this repo at the upstream repo: https://github.com/react-native-clipboard/clipboard

**Comment by roryabraham** [comment]
No new proposals here.

**Comment by puneetlath** [comment]
Upwork job created here: https://www.upwork.com/jobs/~011432fcc2168828e1

@Drewfergusson would you be interested in this? Given you implemented the first part.

**Comment by Drewfergusson** [comment]
@puneetlath things have gotten quite a bit busier for me lately so I'm going to have to pass.

**Comment by MelvinBot** [comment]
@puneetlath, @roryabraham Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by roryabraham** [comment]
No new proposals here

**Comment by MelvinBot** [comment]
@puneetlath, @roryabraham Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
@puneetlath, @roryabraham it looks like no one is assigned to work on this job.
Please double the price or add a comment stating why the job isn't being doubled.

**Comment by puneetlath** [comment]
Updated price to $2000.

**Comment by MTN718** [comment]
# Proposal 
Currently used setString function for copy to clipboard for attachement
![image](https://user-images.githubusercontent.com/12084368/130594192-54db9146-f704-41ec-8c89-34a842887a75.png)

And when paste, it is not parsing content correctly in handlePaste for textinput

![image](https://user-images.githubusercontent.com/12084368/130594391-c85710d1-def0-498f-ba0f-78afabfb2cb1.png)

embeddedImages length always showing 0

Reason: setString to clipboard making extra span tag image element converted to string

To fix issue 
- Get image element text from clipboard
![image](https://user-images.githubusercontent.com/12084368/130594728-3351fdb3-f6af-463f-953d-6e810fe0d8d6.png)

- Convert it to html tag from string
![image](https://user-images.githubusercontent.com/12084368/130594818-52fedaad-70f7-41fd-aed7-8c01bf393d45.png)

- Attached Working Video


https://user-images.githubusercontent.com/12084368/130595083-442b227f-a1e7-4a9e-b71e-7b7304144234.mov


Please review and give me feedback @roryabraham , @puneetlath

**Comment by roryabraham** [comment]
@MTN718

> Reason: setString to clipboard making extra span tag image element converted to string

Sorry, I'm having difficulty understanding this. Can you help me understand this better?

Other than that, there are two main concerns I have with your proposal:

1. Keep in mind that this is a react-native codebase, so we need a solution that works not only on web, but also on iOS and Android. Ideally, we would create a solution that works for both iOS/Android and react-native-web, so that we don't need two separate implementations. See the following resources:
    - [Expensify/App Philosophy](https://github.com/Expensify/App#Philosophy), particularly section 5 titled "Cross Platform 99.9999%"
    - https://github.com/react-native-clipboard/clipboard/issues/6
    - https://github.com/Expensify/App/issues/4629#issuecomment-897981246
1. Also, can you verify with your existing solution that the image binary is copied to the clipboard such that it can be pasted into another program, such as Notes, Google Docs, or Microsoft Word?

**Comment by anboia** [comment]
### Proposal
---

Use native share workflow for mobile (with [react-native-share](https://github.com/react-native-share/react-native-share) ), and keep web/desktop version with the "Copy to Clipboard" Context menu.

---

Explanation:
UX is increased, by having more options: with just one step users can copy the content, or send it to other apps like WhatsApp, Slack, Note (or any other workflow they prefer)

---

Solution

Create a `<Sharable/>` component, similar to `<PressableWithSecondaryInteraction/>`, that instead of showing the context menu, it would show the native ios/android share menu. This is the same behavior that slack takes when long-pressing a link for example:


<img src="https://user-images.githubusercontent.com/4477665/130674883-c01e4852-ff44-49d0-a5ec-78aea92ea10a.png" width="200" />

This is how it would look like in android and ios:

<img src="https://user-images.githubusercontent.com/4477665/130674206-6364ea77-c0af-4551-b683-d10405a49a26.png" width="200" />

<img src="https://user-images.githubusercontent.com/4477665/130674230-4d613880-4c96-4e91-a427-917b0777300c.png" width="200" />

**Comment by roryabraham** [comment]
@anboia That _is_ an interesting alternate solution, I love it!

That said, I think it should really be a separate context menu option. I recommend that you create a separate issue in this repo to propose adding a share option to the context menu using the native share menu, and we can discuss over there. Plus, if we decide to move forward with this new feature, there's an additional bonus we pay to contributors if you open an issue and also implement the solution. 🤑

**Comment by MTN718** [comment]
Found solution for web version.

https://user-images.githubusercontent.com/12084368/130681105-ba4cd577-7988-4c94-8b17-a9d10459381e.mov

Used clipboard.write instead of setString

**Comment by MTN718** [comment]
@roryabraham one question, currently how to open context menu in the ios?

**Comment by MTN718** [comment]
It's mobile web version

https://user-images.githubusercontent.com/12084368/130682485-44ac9a44-5f73-403d-b8f7-f53d76b67f74.mov

**Comment by roryabraham** [comment]
> @roryabraham one question, currently how to open context menu in the ios? long press on image is working?

Yeah, longPress should open the context menu.

**Comment by roryabraham** [comment]
> One Notice: Image url should be png file. clipboard support only image/png type. can't save jpg file to clipboard
So we should be discuss about convert image jpg to png from server or frontend.

Really? I've never encountered this issue

**Comment by parasharrajat** [comment]
@roryabraham I think we already have an open issue #1294 for share menus.

**Comment by roryabraham** [comment]
@parasharrajat https://github.com/Expensify/App/issues/1294 is slightly different – I believe that's about adding New Expensify as an option in native share menus, not about exposing native share menus from within New Expensify

**Comment by parasharrajat** [comment]
Oh yeah. :+1:

**Comment by anboia** [comment]
> One Notice: Image url should be png file. clipboard support only image/png type. can't save jpg file to clipboard
> So we should be discuss about convert image jpg to png from server or frontend.
> 

It seems you should use [data URLs shema](https://datatracker.ietf.org/doc/html/rfc2397 ) with base64, which will work with a variety of `mediatype`, probably all attachment types accepted by the backend API.
We should use the same approach (share the same function) for retriveing the url of the attachment to a base64 data url, so that it is possible to copy or share the base64 data. Here it is a good example on how to do this:
(that would be your `fetch(embeddedImages[0].src)`)
```javascript
.fetch('GET', fileUrl)
    .then(resp => {
      filePath = resp.path();
      return resp.readFile('base64');
    })
    .then(async base64Data => {
      base64Data = `data:${type};base64,` + base64Data;

```
the full example is here: https://react-native-share.github.io/react-native-share/docs/share-remote-file


Also `react-native-web` uses the same `Clipboard` component from `react-native` which is deprecated (https://reactnative.dev/docs/clipboard), so I would recommend using only https://github.com/react-native-clipboard/clipboard for that (rename `src/libs/Clipboard/index.native.js` to `src/libs/Clipboard/index.js`)

**Comment by MTN718** [comment]
@roryabraham I tried long press on ios simulator for image but can't see open menu. 
And can you provide png url for image so i can make it working image copy to clipboard as you can see video above.

**Comment by MTN718** [comment]
@anboia i know that solution. but if already there is png image in the server, no need to convert it to using base64. so @roryabraham I want discuss with you about convert in frontend or just use png url from server

**Comment by MTN718** [comment]
Any feedback for my proposal?

**Comment by MTN718** [comment]
@roryabraham this issue still opening? no update yet

**Comment by roryabraham** [comment]
@MTN718 Very sorry for the delay here. Your last proposal looks fantastic! @puneetlath Let's hire @MTN718 for this job on Upwork!

**Comment by MTN718** [comment]
Thanks @roryabraham First I need to use expensify forked library for customize?

**Comment by roryabraham** [comment]
@MTN718 Check out [this comment above](https://github.com/Expensify/App/issues/4629#issuecomment-897981246). While you're developing locally, you should work off of your own forked copy of https://github.com/react-native-clipboard/clipboard/. Then once you're confident that it's working well and can show us some of your local testing results, then you can create a PR from your forked copy of https://github.com/react-native-clipboard/clipboard/ to ours. Does that makes sense? The comment above explains in more detail.

**Comment by roryabraham** [comment]
In the meantime, I'll create an internal issue to create an Expensify/clipboard fork.

**Comment by roryabraham** [comment]
Just FYI, I just noticed [this PR](https://github.com/react-native-clipboard/clipboard/pull/112), which might implement image-copying in react-native-clipboard. We will still hire you to ensure that we get this feature implemented in Expensify/App, but how similar is that PR to your current implementation?

**Comment by MTN718** [comment]
In my implementation, input parameter is Image URL but in current PR, not. And in my view, that's not complete solution.

**Comment by MTN718** [comment]
@roryabraham In android, copy paste image support for some android OS i think. 

![image](https://user-images.githubusercontent.com/12084368/131715673-eb0880fa-644a-4dbc-8a14-e48a5b7f632f.png)

**Comment by roryabraham** [comment]
Interesting, how up-to-date is that StackOverflow? cc @Julesssss Do you know anything about Android copy/paste capabilities?

**Comment by MTN718** [comment]
@roryabraham one thing more, when request fetch of image data from website (local testing), it shows cors policy error because i'm testing local and fetch from expensify.com. It will be not problem when testing on your side correct?

**Comment by MTN718** [comment]
and still not hired yet on upwork

**Comment by roryabraham** [comment]
> @roryabraham one thing more, when request fetch of image data from website (local testing), it shows cors policy error because i'm testing local and fetch from expensify.com. It will be not problem when testing on your side correct?

While I don't think this will be a problem on our side, can you please provide a screenshot of the console errors your seeing so we can see about eliminating this issue for outside contributors in the future. Are you using the web proxy provided in the repo? (i.e: can you sign into an Expensify account while testing on your local branch?)

> still not hired yet on upwork

Bump @puneetlath

**Comment by roryabraham** [comment]
Oh yeah, and we've forked react-native-clipboard: https://github.com/Expensify/Clipboard

**Comment by MTN718** [comment]
![image](https://user-images.githubusercontent.com/12084368/131724614-3c9da7fa-69e3-42ee-879d-b82155238c75.png)

This is console error when fetch image data

**Comment by roryabraham** [comment]
Thanks. Oh, and please try to do most of your testing with a test account instead of Concierge. I believe test messages sent to Concierge may clutter our customer support channels.

**Comment by MTN718** [comment]
Now need to add new branch on your forked react native clipboard library?

**Comment by roryabraham** [comment]
You should be able to create a pull request from your forked react native clipboard library to ours

**Comment by MTN718** [comment]
https://github.com/Expensify/clipboard/pull/1
Created PR

**Comment by roryabraham** [comment]
Okay, I'll review as soon as I can.

**Comment by Julesssss** [comment]
> Interesting, how up-to-date is that StackOverflow? cc @Julesssss Do you know anything about Android copy/paste capabilities?

Sorry, just catching up here.

Honestly copying an image makes no sense to me as a mobile user. I can't find any apps that let you copy and paste an image and I don't think we should add this to our app (even if we did, I doubt any app would support the pasted content anyway). The closest equivalent action for mobile should be 'save image', as the user can share it to any app which supports image sharing.

I'm not as familiar with iOS, but it seems to be exactly the same from my testing.

**Comment by Julesssss** [comment]
I [left a review](https://github.com/Expensify/clipboard/pull/1#pullrequestreview-746006688) on the PR with some additional information.

**Comment by roryabraham** [comment]
> Honestly copying an image makes no sense to me as a mobile user. I can't find any apps that let you copy and paste an image

This isn't my experience on iOS – I use native copy/paste functionality all the time in Messenger, Facebook, WhatsApp, etc...

|iMessage|WhatsApp|Slack|Facebook Messenger|
|--------|--------|-----|------------------|
|![iMessage](https://user-images.githubusercontent.com/47436092/132004200-c645f132-1340-434a-bba6-df72cc4b8cd3.jpeg)|![WhatsApp](https://user-images.githubusercontent.com/47436092/132004168-2b1e229b-0aaa-4ce9-913f-f253720b6655.jpeg)|![Slack](https://user-images.githubusercontent.com/47436092/132004205-63c2e4cb-9ba6-4ed7-bac9-afffe394795a.jpeg)|![Facebook Messenger](https://user-images.githubusercontent.com/47436092/132004210-a0280ce5-6899-41d3-87db-f20d98d77f7f.jpeg)|

Overall I would saw copying base64-encoded images to the native clipboard is a reasonably common operation, at least on iOS.

> even if we did [implement native image-copying], I doubt any app would support the pasted content anyway

I'm not convinced this will be a problem – if image-copying works smoothly between all of the above-listed apps, why not ours as well? I think base64-encoded images are pretty standard things that most apps know how to process when pasted.

Another reason I support implementing this functionality is to preserve cross-platform parity per the Expensify/App Philosophy of [_Cross Platform 99.9999%_](https://github.com/Expensify/App#Philosophy). If our UI enables an action (such as image-copying) on web/desktop, it should enable the same action on native platforms. The only exception to this parity is when an action relies on some hardware (such as a mouse) or system component (such as native share menus) for which there is no equivalent on the native platforms. However, as far as I can tell that's just not the case with clipboard.

---

**Note:** To be clear I do absolutely agree that we should expose native share menus to our app UI and register our app as a receiver of images via native share. But imo these three are all separate issues:

- Right now, our app does not support copying of images on through the UI, nor does it support copying of images on native platforms at all.
- We currently have no way to share an image to arbitrary other apps via the native share menu,
- New Expensify does not appear in the native share menu as a potential recipient. (although our legacy app actually _does_ show up, and sharing via that begins a SmartScan of the image)

**Comment by Julesssss** [comment]
Ah weird, I did check iOS Messenger & WhatsApp and didn't see the copy image option in either of the apps?! Maybe because I have an older iPhone? (iPhone 7, 13.3) 😕  But fair enough, this must be a common iOS operation then.

I still feel that this is not worth implementing for Android though. Forking and extending the library for feature purely is a waste of time in this case as image copy/paste isn't supported by the OS (as MTN [shared here](https://github.com/Expensify/App/issues/4629#issuecomment-910493163)). I don't agree that we should implement a hacky single platform feature simply to claim 99.99% alignment. However, as the feature is already completed then maybe it should be included 🤷‍♂️ 

Also I do agree with our cross-platform philosophy, but we already have exceptions to this rule: [Android doesn't support notification badge count like iOS], [desktop doesn't let you take photos's like mobile].

**Comment by MTN718** [comment]
Yes IOS copy base64 image to clipboard No problem. But Android not support is problem on our project. How to resolve it?

**Comment by roryabraham** [comment]
We've been working with @MTN718 in the PR against `Expensify/Clipboard` – hit a snag in that the current implementation won't work offline.

**Comment by kadiealexander** [comment]
Hi @MTN718, I've placed this issue on hold as per [this update](https://expensify.slack.com/archives/C01GTK53T8Q/p1631754482374800), as we are prioritising issues related to a feature release scheduled for 9/24. As an apology for the delay, we will add a $100 bonus as a thank you for waiting.

**Comment by roryabraham** [comment]
Still on hold

**Comment by MTN718** [comment]
still hold issue?

**Comment by kadiealexander** [comment]
Please refer to [this post](https://expensify.slack.com/archives/C01GTK53T8Q/p1633557259458600) for updated information on the `n6 hold`, we've raised the bonus to **$250** for all issues where a **PR is created** before the N6 hold is lifted.

**Comment by roryabraham** [comment]
Taking this off hold, but I have a large backlog of pending reviews so it may be a couple days before I get to this.

**Comment by MTN718** [comment]
Let me know when need to start

**Comment by puneetlath** [comment]
@roryabraham @MTN718 how's this going?

**Comment by MTN718** [comment]
I'm waiting rory's confirmation.

**Comment by mallenexpensify** [comment]
We are placing all non-critical issues on hold while we're on a company offsite.  The hold is expected to be lifted on 11/19 (but could go longer).  For any PRs that are submitted before or during the hold, we will add a $250 bonus payment.

**Comment by mateomorris** [comment]
# Proposal
Update the `Copy to Clipboard` item in `ContextMenuActions`

Given the image url (which can be extracted from `html`), the image needs to be converted into a blob before being copied to the clipboard.
![image](https://user-images.githubusercontent.com/39444813/140812486-3d4a7555-b423-4b6f-ba9d-4291bb60ceb2.png)

Normally, we might redownload it to convert it into a blob, but that wouldn’t work offline. If we’re offline, we need to access the node itself to get the image data (will change this to use a ref). 
![image](https://user-images.githubusercontent.com/39444813/140812709-0dc0792a-ef1a-4cc9-b5fa-8a1d8c337b36.png)

Once we have the image data and have converted it into a blob, we can add it the clipboard (will look for a solution on iOS & android). 

![image](https://user-images.githubusercontent.com/39444813/140812032-9f4ba5d4-bbf7-4102-9efb-423321d4a63b.png)

**Comment by roryabraham** [comment]
> Another option for getting the image data that doesn't depend on access to the node (which might be complicated on mobile, not sure yet) is to set the actual src on the image to a data uri (e.g. base64 string), and then converting that base64 string into a blob which gets added to the clipboard.

This is a promising thought. I don't think we'll want to render the image on a canvas just to copy it.

**Comment by mateomorris** [comment]
@roryabraham agreed. Then using a datauri is the only solution that comes to mind. Apologies for the delay on finishing out my proposal, been having a lot of trouble lately running mobile locally.

**Comment by puneetlath** [comment]
Doubling price to $4000: https://www.upwork.com/jobs/~012fd9a807faaf42aa

**Comment by MTN718** [comment]
@roryabraham to copy image data from element, need to change chat part also. Because currently we're using image from uri in the chat.

**Comment by roryabraham** [comment]
Curious to see a proposal that implements images + image coping using a base64 data uri. I'm sure there could be a number of _gotchas_ we'll need to be wary of.

**Comment by roryabraham** [comment]
> @roryabraham actually, I'm working on another solution that utilizes an existing PR for offline images. Just testing it before updating my proposal.

Exciting, can't wait!

**Comment by mateomorris** [comment]
@roryabraham never mind, the PR turned out not to be helpful in this case as far as I can tell. 

Would you mind elaborating a bit on the criteria for the solution here? It looked like @MTN718's proposal would still work for instances where the image is cached on the page (at least it did for me, locally). I assume we wouldn't need to copy an image if it isn't rendering on the page, so the challenge is to copy the visible image to the clipboard either by fetching it like normal (ideally getting a cached response back), or retrieving and parsing the data out of the DOM node (with one of the methods I outlined in my proposal or another I'm overlooking). 

Is that right?

**Comment by roryabraham** [comment]
> Would you mind elaborating a bit on the criteria for the solution here?

Yep, the solution needs to:

1. Work on web, mobile web, iOS, Android, and desktop.
2. Not be dependent upon a network connection.

> It looked like @MTN718's proposal would still work for instances where the image is cached on the page (at least it did for me, locally).

Does that work for iOS/Android when the app is offline?

**Comment by parasharrajat** [comment]
@roryabraham Correct, we need to make images offline first so that we can refer to them in the file system. I think https://github.com/Expensify/react-native-onyx/pull/102 will enable us to do that easily so I think we should hold this issue till that time. 
Until we store the images in APP storage (file system) there is no way to get them without the internet. Currently, they are only located via a URL so first have a feature to make them offline, and then we can easily use the path as a URL and add a copy feature.
IMO, this feature can't be offline until we have offline images. Thus I think we should hold this.

**Comment by puneetlath** [comment]
Thanks @parasharrajat I've removed the help wanted label and marked this as on hold for now.

**Comment by puneetlath** [comment]
Still on hold.

**Comment by puneetlath** [comment]
@roryabraham @parasharrajat now that https://github.com/Expensify/react-native-onyx/pull/102 has been merged, can we reopen this issue?

**Comment by roryabraham** [comment]
@puneetlath I think we need offline images in the app first. https://github.com/Expensify/react-native-onyx/pull/102 laid the groundwork for that, but we will still need to implement non-URL-based images in New Expensify before we can worry about copying those images to the clipboard.

**Comment by puneetlath** [comment]
Ok got it! And do we have an issue for implementing offline images somewhere?

**Comment by roryabraham** [comment]
I asked, and it sounds like @thienlnam has a good handle on the next steps for offline images, and gave some additional context [over here](https://github.com/Expensify/App/issues/3867#issuecomment-996233570)

**Comment by parasharrajat** [comment]
After we have added functionality to add images offline in NewDot. We will need to make attachment offline.

For that, there is a issue https://github.com/Expensify/App/issues/6527

**Comment by roryabraham** [comment]
Order looks correct - this is still on hold for now. I'm going to make it a monthly while it's still on hold.

**Comment by roryabraham** [comment]
I believe this is still on HOLD pending https://github.com/Expensify/App/issues/6527

**Comment by roryabraham** [comment]
Still on HOLD

**Comment by roryabraham** [comment]
Still on HOLD

**Comment by roryabraham** [comment]
This is on hold for an issue that's not being actively worked on.... https://github.com/Expensify/App/issues/6527

**Comment by roryabraham** [comment]
still on HOLD

**Comment by roryabraham** [comment]
Still on HOLD

**Comment by roryabraham** [comment]
Still on HOLD for an issue that's in planning. Adding the planning label

**Comment by roryabraham** [comment]
Still on HOLD, reassigning to @Beamanator

**Comment by Beamanator** [comment]
Ya I think this should hold on caching, looks like we're close so it should only hold another week!

**Comment by Beamanator** [comment]
Still on hold, barely

**Comment by Beamanator** [comment]
Still on hold

**Comment by Beamanator** [comment]
Still held on caching

**Comment by Beamanator** [comment]
Stillllll held

**Comment by Beamanator** [comment]
still holding

**Comment by trjExpensify** [comment]
Still blocked on the server migration project dependency.

**Comment by trjExpensify** [comment]
No change.

**Comment by trjExpensify** [comment]
Getting closer! CORS issue resolved, web image caching PR in development. :)

**Comment by trjExpensify** [comment]
Image caching PR was deployed and reverted. Awaiting this now before re-implementing: https://github.com/Expensify/App/issues/32703

**Comment by blimpich** [comment]
https://github.com/Expensify/App/issues/32703 is fixed now.

**Comment by trjExpensify** [comment]
In light of #32703 being fixed, the [new image caching PR](https://github.com/Expensify/App/pull/34505) is in review.

**Comment by trjExpensify** [comment]
The image caching PR was [reverted again](https://github.com/Expensify/App/pull/35636) on Friday, so we're back to figuring that out in #12603.

**Comment by trjExpensify** [comment]
This is the latest PR: https://github.com/Expensify/App/pull/36560

**Comment by trjExpensify** [comment]
Okay, that has merged. Can we pick this up again, or do we need something else first @Beamanator @roryabraham?

**Comment by trjExpensify** [comment]
Also, curious if we have any overlap now with @kidroca's ongoing work with image/attachments improvements.

**Comment by trjExpensify** [comment]
>Though I think we need to be clear: Do we still agree that this feature isn't required for Android? We argued that a few years back but maybe now things are different?

Does Android now support copying images like iOS does on more recent OS versions?

**Comment by Beamanator** [comment]
That's definitely a good question, I don't have my android phone with me this week so I can't test :/ but I can ask in slack!

**Comment by trjExpensify** [comment]
Any luck with that, Alex?

**Comment by trjExpensify** [comment]
That would make sense to me.

**Comment by Beamanator** [comment]
Cool so how about we open this back up for proposals! 🚀

**Comment by trjExpensify** [comment]
@roryabraham are you going to close this in favour of https://github.com/Expensify/App/issues/41239 or something? 😕

**Comment by Beamanator** [comment]
I'm fine with moving this to that new issue as long as we align on the desired outcome 👍

**Comment by Beamanator** [comment]
There's a decent amount of activity in https://github.com/Expensify/App/issues/41239 so I'm going to close this one for that one
